src/codemirror/extensions/autocomplete.ts

```diff
@@ -0,0 +1,180 @@
+): Completion[] {
+  return Object.keys(options)
+    .filter((componentName) =>
+      componentName.toLowerCase().startsWith(partialName.toLowerCase())
+    )
+    .map((componentName) => ({
+      label: componentName,
+      section: "Components",
+      type: "keyword",
+      apply: (view, completion, from, to) => {
+        // insert the component name and a space
+        const insertText = `${completion.label} `;
+        view.dispatch({
+          changes: { from, to, insert: insertText },
+          // move the cursor to the end of the new inserted text
+          selection: { anchor: from + insertText.length },
+        });
+
+        // we need to trigger autocompletion to immediately show prop completions after inserting the component name
+        requestAnimationFrame(() => {
+          startCompletion(view);
+        });
+      },
+    }));
+}
+
+function extractAlreadyUsedProps(
+  fullLineText: string,
+  cursorPos: number
+): Set<string> {
+  const tagStart = fullLineText.lastIndexOf("<", cursorPos);
+  let tagEnd = fullLineText.indexOf(">", cursorPos);
+  if (tagEnd === -1) {
+    // tags that aren't closed yet
+    tagEnd = fullLineText.length;
+  }
+
+  const tagContent = fullLineText.substring(tagStart, tagEnd);
+
+  // props in the format propName="value", propName={value}, and just propName (flag-boolean)
+  const propRegex = /(\w+)(?=\s*=|\s*\/>|>|$)/g;
+  const usedProps = new Set<string>();
+  let match;
+
+  while ((match = propRegex.exec(tagContent))) {
+    usedProps.add(match[1]);
+  }
+
+  return usedProps;
+}
+
+function generatePropCompletions(
+  componentName: keyof AutocompletionsMetadata,
+  options: AutocompletionsMetadata,
+  partialPropName: string,
+  usedProps: Set<string>,
+  lineTextUpToCursor: string
+): Completion[] {
+  const componentProps = options[componentName];
+  if (!componentProps?.length) {
+    return [];
+  }
+
+  const normalizedProps = componentProps.flatMap((propItem) =>
+    Object.entries(propItem).map(([name, propInfo]) => ({ name, ...propInfo }))
+  );
+
+  return normalizedProps
+    .filter(
+      ({ name }) => !usedProps.has(name) && name.startsWith(partialPropName)
+    )
+    .map((item) => {
+      console.log({ item });
+      return item;
+    })
+    .map(({ name, type, required, defaultValue, description }) => ({
+      label: name,
+      boost: required ? 1 : 0,
+      detail: Array.isArray(type) ? type.join(" | ") : type,
+      info:
+        (description ? `${description}` : "") +
+        (description && defaultValue ? " | " : "") +
+        (defaultValue ? ` Defaults to: ${defaultValue}` : ""),
+      section: `${componentName}'s props`,
+      type: required ? "required" : "property",
+      apply: (view, _completion, from, to) => {
+        const applyText = type === "string" ? `${name}=""` : `${name}={}`;
+        const applyPrefix = lineTextUpToCursor.endsWith(" ") ? "" : " ";
+        const textToInsert = applyPrefix + applyText;
```

Comment:
Same

Reply:
so in here I wanted to make the dx better by applying automatically `""` or `{}` based on the prop type
for example when typing `<Button dis` and choosing a boolean prop called `disabled` from the list it will autocomplete to `disabled={}`.
when typing `<Button aria-la` and choosing a string prop called `aria-label` from the list it will autocomplete to `aria-label=""`.

and the "prefix" spacing is done also for dx purposes, to not cause a situation where if someone adds space it adds another space, or having a prop string attached to the previous prop or component tag

Reply:
basically from all the manual QA I did to this it looks fine, but even if it the logic would fail for some reason, it isn't crucial as it doesn't affect the output severely

I do need to add a lot of tests for all this extension, though